superset/viz.py

```diff
@@ -202,7 +222,7 @@ def query_obj(self):
         """Building a query object"""
         form_data = self.form_data
         gb = form_data.get('groupby') or []
-        metrics = form_data.get('metrics') or []
+        metrics = self.all_metrics or []
```

Comment:
I don't like the pattern we currently have for returning empty arrays, i feel like we should be throwing some exceptions with the proper messaging to let the user know how that there are no metrics currently selected.

Reply:
The UI does a pretty good job right now informing users that they do not have a metric selected and preventing them from issuing queries. I'm not sure what situations there could be where a user could submit a query from explore without a metric selected, but I am all for informative error messages. I still think its good to have an idea of when this could happen so the error message offers the right information. Is this if there's a bug in the UI? Or are there cases where the UI is working and someone still can press `Run Query` with no metric selected? See:

https://github.com/apache/incubator-superset/blob/master/superset/assets/src/explore/controls.jsx#L176

Reply:
cc @hughhhh @mistercrunch ^

Reply:
It's a bit complex here as many controls may use `MetricControl` (some with non null validators, some without) and we don't want to duplicate logic. In the context above `metrics` may exist as a key or not exist, or have `None`, `[]` or an actual list of values OR dicts. That first step here is to remove `None` or key doesn't exist from the list of possibility as it should be treated the same.

There's clearly much that could be done with a stronger contract between the backend and the frontend.